donkeycar/pipeline/training.py

```diff
@@ -0,0 +1,112 @@
+import math
+import os
+from pathlib import Path
+from typing import List
+
+from donkeycar.parts.tflite import keras_model_to_tflite
+from donkeycar.pipeline.sequence import TubRecord
+from donkeycar.pipeline.sequence import TubSequence
+from donkeycar.pipeline.types import TubDataset
+from donkeycar.pipeline.augmentations import ImageAugmentation
+from donkeycar.utils import get_model_by_type, normalize_image
+import tensorflow as tf
+
+
+class BatchSequence(object):
+    """
+    The idea is to have a shallow sequence with types that can hydrate
+    themselves to np.ndarray initially and later into the types required by
+    tf.data (i.e. dictionaries or np.ndarrays).
+    """
+    def __init__(self, model, config, records: List[TubRecord], is_train: bool):
+        self.model = model
+        self.config = config
+        self.sequence = TubSequence(records)
+        self.batch_size = self.config.BATCH_SIZE
+        self.is_train = is_train
+        self.augmentation = ImageAugmentation(config)
+        self.pipeline = self._make_pipeline()
+
+    def __len__(self):
+        return math.ceil(len(self.pipeline) / self.batch_size)
+
+    def _make_pipeline(self):
+        """ This can be overridden if more complicated pipelines are
+            required """
+        # 1. Initialise TubRecord -> x, y transformations
+        def get_x(record):
+            x0 = self.model.x_transform(record)
+            x1 = self.augmentation.augment(x0) if self.is_train else x0
+            x2 = normalize_image(x1)
+            return x2
+
+        def get_y(record):
+            return self.model.y_transform(record)
+
+        # 2. Build pipeline using the transformations
+        pipeline = self.sequence.build_pipeline(x_transform=get_x,
+                                                y_transform=get_y)
+
+        return pipeline
+
+    def make_tf_data(self):
+        dataset = tf.data.Dataset.from_generator(
+            generator=lambda: self.pipeline,
+            output_types=self.model.output_types(),
+            output_shapes=self.model.output_shapes())
+        return dataset.repeat().batch(self.batch_size)
```

Comment:
Why `repeat()`  with no args ?

You might want to specify a `count` of the times it can be repeated. It should be set to `1` (otherwise we might end up overfitting).  The docs (https://www.tensorflow.org/api_docs/python/tf/data/Dataset#repeat) say:

```

(Optional.) A tf.int64 scalar tf.Tensor, representing the number of times the dataset should be repeated. The default behavior (if count is None or -1) is for the dataset be repeated indefinitely.
--

That's not what we want right ?

Reply:
If we set this to `1` then we can run through the data only once, i.e. only one epoch. After the first epoch we are getting: 
```
WARNING:tensorflow:Your input ran out of data; interrupting training. Make sure that your dataset or generator can generate at least `steps_per_epoch * epochs` batches (in this case, 780 batches). You may need to use the repeat() function when building your dataset.
```